errbot/flow.py

```diff
@@ -0,0 +1,335 @@
+
+    def next_autosteps(self) -> List[FlowNode]:
+        """
+        Get the next steps that can be automatically executed according to the set predicates.
+        """
+        return [node for predicate, node in self._current_step.children if predicate(self.ctx)]
+
+    def next_steps(self) -> List[FlowNode]:
+        """
+        Get all the possible next steps after this one (predicates statisfied or not).
+        """
+        return [node for predicate, node in self._current_step.children]
+
+    def advance(self, next_step: FlowNode, enforce_predicate=True):
+        """
+        Move on along the flow.
+        :param next_step: Which node you want to move the flow forward to.
+        :param enforce_predicate: Do you want to check if the predicate is verified for this step or not.
+                                   Usually, if it is a manual step, the predicate is irrelevant because the user
+                                   will give the missing information as parameters to the command.
+        """
+        if enforce_predicate:
+            predicate = self._current_step.predicate_for_node(next_step)
+            if predicate is None:
+                raise ValueError("There is no such children: %s" % next_step)
+
+            if not predicate(self.ctx):
+                raise InvalidState("It is not possible to advance to this step because its predicate is false")
+
+        self._current_step = next_step
+
+    @property
+    def name(self):
+        """
+        Helper property to get the name of the flow.
+        """
+        return self._root.name
+
+    def __str__(self):
+        return "%s (%s) with params %s" % (self._root, self.requestor, dict(self.ctx))
+
+
+class BotFlow(IPlugin):
+    """
+    Defines a Flow plugin ie. a plugin that will define new flows from its methods with the @botflow decorator.
+    """
+    def __init__(self, bot):
+        super().__init__()
+        self._bot = bot
+        self.is_activated = False
+
+    def activate(self) -> None:
+        """
+            Override if you want to do something at initialization phase (don't forget to
+            super(Gnagna, self).activate())
+        """
+        self._bot.inject_flows_from(self)
+        self.is_activated = True
+
+    def deactivate(self) -> None:
+        """
+            Override if you want to do something at tear down phase (don't forget to super(Gnagna, self).deactivate())
+        """
+        self._bot.remove_flows_from(self)
+        self.is_activated = False
+
+    def get_command(self, command_name: str):
+        """
+            Helper to get a specific command.
+        """
+        self._bot.all_commands.get(command_name, None)
+
+
+class FlowExecutor(object):
+    """
+    This is a instance that can monitor and execute flow instances.
+    """
+    def __init__(self, bot):
+        self._lock = RLock()
+        self.flow_roots = {}
+        self.in_flight = []
+        self._pool = ThreadPool(5)
```

Comment:
I gonna extract that to an external constant.
This is true only for flows currently executing in automatic mode, once it has nothing more, it will free up the thread.
so 5 concurrent automatic workflows at the same time is already a lot. If it causes a contention for our users will add that as a config parameter.